### ISSUE TITLE
Matrix norm does not seem implemented for order different from 2 contrarily to what doc says

### ISSUE THREAD

**Issue by mick-d** [numpy/numpy#10705]
When using `numpy.linalg.norm` with `ord` different from 2, for example 3, then I expected the norm to be computed according to the documentation, i.e. as `sum(abs(x)**ord)**(1./ord)`

But instead I got:
```
ValueError                                Traceback (most recent call last)
<ipython-input-784-3344f4d8d663> in <module>()
----> 1 nnorm((distpdfs_diverg['BC']/n_samples) - np.eye(n_bundles), 3)

/usr/lib/python3.5/site-packages/numpy/linalg/linalg.py in norm(x, ord, axis, keepdims)
   2247             ret = _multi_svd_norm(x, row_axis, col_axis, sum)
   2248         else:
-> 2249             raise ValueError("Invalid norm order for matrices.")
   2250         if keepdims:
   2251             ret_shape = list(x.shape)

ValueError: Invalid norm order for matrices.
```
Looking at the source code (numpy 1.13, numpy 1.14, and [current](https://github.com/numpy/numpy/blob/e3fe42f0473d41f97db5a40e3ddd067391df57f0/numpy/linalg/linalg.py#L2132)) it seems it is simply the norm is simply not implemented for `ord` greater than 2 contrarily to what the doc says. Would that be something difficult to do?

**Comment by eric-wieser** [comment]
This is as documented:

    =====  ============================  ==========================
    ord    norm for matrices             norm for vectors
    =====  ============================  ==========================
    other  --                            sum(abs(x)**ord)**(1./ord)
    =====  ============================  ==========================

Note the column for matrices is blank

**Comment by mick-d** [comment]
Indeed, my bad... 

But then what about implementing the <ord>-norm for (n x m) matrices? 
`sum[i=1..n][j=1..m](abs(a_ij)**ord)**(1./ord)`

**Comment by pv** [comment]
That does not agree with the definition of these matrix norms. The `p` 
norm of `A` of size `n \times m` is *defined* as

`norm_p(A) := \sup_{x \in \mathbb{C}^m} ||A x||_p / ||x||_p`

where `||.||_p` is the corresponding vector norm. Unfortunately, in 
practice this is hard to compute, but that's the standard definition.

**Comment by mick-d** [comment]
@pv I meant the L_{p,p} norm, i.e. the simple extension of the Frobenius norm for which p=2 in the formula i wrote above (`ord`=p). Intuitively, it feels it should not be very intensive to compute but my intuition may be wrong...

**Comment by pv** [comment]
No, numpy.linalg.norm(A, ord=2) returns the matrix 2-norm, not the 
Frobenius norm. So it's not consistent to have `norm(A, ord=3)` 
completely different norm...

**Comment by mick-d** [comment]
Damn it, I keep making mistakes today, you are totally right @pv.

Still, would it be possible to implement the L(p,p) norm? The silver lining of being totally wrong about `ord` having something to do with this, is that it may be even easier to implement the L_{p,q} norm to generalize the Frobenius norm? 

As a (totally random) call example, something like `numpy.linalg.norm(A, L=(p,q))`

**Comment by pv** [comment]
In principle, we could imagine adding `ord=("fro", n)` for the 
Frobenius-like norms. But I'm not sure if that makes sense --- if you 
need those, you can as well write `norm(A.ravel(), ord=n)`.

The error message for not implemented matrix norms could be more clear, 
however.

**Comment by mick-d** [comment]
I think it would be great to have some generalization of the Frobenius norm, which is just a special case of the L_{p,q} norm in reality. L_{p,p} would be great and maybe even L_{p,q} if this is quite similar in implementation to L_{p,p}. I am not a numerical analyst but i often encounter the L norm in math problems.

PS: I think the error message was pretty good!

**Comment by mick-d** [comment]
Excellent tip for the implementation... it is true that L_{p,p} is really just ` norm(A.ravel(), ord=p)`. The question would just to know if implementing L_{p,q} is worth the effort. Maybe it is not the case.

**Comment by pv** [comment]
Yes, `ord=(p, q)` could also be possible. The implementation is in 
principle one-liner.

**Comment by mick-d** [comment]
Great news, then i personally believe this could be a great addition. Not sure what others think.

**Comment by mhvk** [comment]
@mick-d - we're hoping to eventually get rid of the matrix subclass -- see #10142 -- as there are many problems with it (not least of which the inability to have stacks of matrices). I would recommend switching to regular arrays and start using the `@` operator...  More relevant to the present topic: perhaps we should have a separate function (or keyword) to get the *matrix* norm given an input *array*.

**Comment by eric-wieser** [comment]
@mhvk: Not sure `np.matrix` is relevant to this discussion.

**Comment by mhvk** [comment]
Am sorry, I should have read more carefully....
